### PR TITLE
add logger

### PR DESCRIPTION
**Pull request by Exortions**
Custom logger for LemLib.

Untested: Colors may or may not work
Todo: Log things in actual LemLib implementations

Due to git issues, it is re-committing the doxygen pr.

### CODE REVIEW

**Comment by liam-teale** [comment]
This should be implemented in existing functions (for example, throwing an error when trying to move the chassis before it has been calibrated). I'll take a closer look this afternoon and do some testing

**Comment by Exortions** [comment]
yeah that was part of the todo, the logger itself is implemented but it needs to be called in functions.

will work on this later

File: src/lemlib/logger.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        case lemlib::logger::Level::ERROR:
+            return "\033[0;31m"; // red
+        case lemlib::logger::Level::FATAL:
+            return "\033[0;31;2m";
+        default:
+            return "\033[0m"; // reset (white)
+    }
+}
+
+std::string getFormattedLevel(lemlib::logger::Level level) {
```
Comment (SizzinSeal) [inline src/lemlib/logger.cpp:96]: Missing break statements after each case

Comment (liam-teale) [inline src/lemlib/logger.cpp:98]: Missing break statements after each case

File: src/lemlib/logger.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+void lemlib::logger::log(Level level, const char* message, const char* exception) 
+{
+    if (!checkLowestLevel(level)) return;
+    if (level == Level::DEBUG && !lemlib::debug) return;
+    if (level == Level::INFO && !lemlib::verbose) return;
+
+    if (message == nullptr) message = "";
+    if (exception == nullptr) throw std::invalid_argument("exception cannot be null");
+
+    std::string messageString = "[LemLib] " + getFormattedLevel(level) +  ": " + message + ": " + exception;
```
Comment (liam-teale) [inline src/lemlib/logger.cpp]: missing reset ansi

File: src/lemlib/logger.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    */
+void lemlib::logger::log(Level level, const char* message) 
+{
+    if (!checkLowestLevel(level)) return;
+    if (level == Level::DEBUG && !lemlib::debug) return;
+    if (level == Level::INFO && !lemlib::verbose) return;
+
+    if (message == nullptr) message = "";
+
+    std::string messageString = "[LemLib] " + getFormattedLevel(level) +  ": " + message;
```
Comment (liam-teale) [inline src/lemlib/logger.cpp]: missing reset ansi

**Review by liam-teale — CHANGES_REQUESTED** [review]

**Pull request merged**